main/server/apps/grid.ts

```diff
@@ -0,0 +1,206 @@
+    staticLocations,
+    gridRootDirectories,
+  }: {
+    name?: AppName;
+    staticLocations: string[];
+    gridRootDirectories: string[];
+  }) {
+    super();
+
+    if (name) {
+      this.name = name;
+    }
+
+    this.staticLocations = staticLocations;
+    this.gridRootDirectories = gridRootDirectories;
+  }
+
+  private async addValidStaticLocations(): Promise<string[]> {
+    const validStaticLocations = [];
+
+    for await (const location of this.staticLocations) {
+      try {
+        await fs.promises.access(location);
+
+        validStaticLocations.push(location);
+      } catch {}
+    }
+
+    return validStaticLocations;
+  }
+
+  /**
+   * Goes through the potential grid roots and looks for sqlite files.
+   * Each Grid 3 root can have any number of users, we return all the users that we find
+   *
+   * A Grid 3 folder has this structure:
+   * /Grid 3
+   *  /Users
+   *    /first
+   *      /en-GB
+   *        /Phrases
+   *          history.sqlite
+   *    /second
+   *      /en-GB
+   *        /Phrases
+   *          history.sqlite
+   */
+  private async addValidDynamicLocations(): Promise<string[]> {
+    const validDynamicLocations: string[] = [];
+
+    for await (const currentRoot of this.gridRootDirectories) {
+      try {
+        const userDirectory = path.join(currentRoot, "./Users");
+
+        await fs.promises.access(userDirectory);
+
+        const allFilesInUserDir = await fs.promises.readdir(userDirectory, {
+          withFileTypes: true,
+        });
+        const users = allFilesInUserDir.filter((source) =>
+          source.isDirectory()
+        );
+
+        for await (const currentUser of users) {
+          const userHistory = path.join(
+            userDirectory,
+            currentUser.name,
+            "./en-GB/Phrases/history.sqlite"
+          );
+
+          await fs.promises.access(userHistory);
+
+          validDynamicLocations.push(userHistory);
+        }
+      } catch {}
+    }
+
+    return validDynamicLocations;
+  }
+
+  private async getLocations(): Promise<string[]> {
+    if (this.validLocations) {
+      return this.validLocations;
+    }
+
+    const validDynamicLocations = await this.addValidDynamicLocations();
+    const validStaticLocations = await this.addValidStaticLocations();
+
+    this.validLocations = [...validDynamicLocations, ...validStaticLocations];
+
+    return this.validLocations;
+  }
+
+  async doesExist() {
+    try {
+      const locations = await this.getLocations();
+      return locations.length > 0;
+    } catch {
+      return false;
+    }
+  }
+
+  /**
+   * We get the raw contents of all the files and hash them.
+   *
+   * We combine the hashes so that if any file changes we
+   * re-fetch the text
+   */
+  async getHash() {
+    const locations = await this.getLocations();
+    const hashes = await Promise.all(
+      locations.map(async (location) => await getHashFromFile(location))
+    );
+    const allHashes = hashes.join("");
```

Comment:
Might want to hash the list of hashes here instead of concatenating them so the return is a consistent length across providers.